app/models/concerns/stat/activity_history.rb

```diff
@@ -2,59 +2,212 @@ class Stat < ApplicationRecord
   module ActivityHistory
     extend ActiveSupport::Concern
 
+    # Example of days stats:
+    # 'days' => {
+    #   '2017' => {
+    #     '08' => {
+    #       '20' => {
+    #
+    #       }
+    #     }
+    #   }
+    # }
+
+    # both dates will be overwritten, just needed the format
     DEFAULT_STATS = {
-      'total' => 0,
-      'activity' => []
+      'days' => {},
+      'first_event_date' => 5.years.ago,
+      'last_update_date' => 5.years.ago,
+      'week_high_score' => 0,
+      'total_progress' => 0
     }.freeze
 
     def recalculate!
-      activity = user.library_events.by_kind(media_column)
-                     .eager_load(media_column)
-
-      # clear stats_data
       self.stats_data = {}
-      self.stats_data = {
-        total: activity.count,
-        activity: activity # collection of events
-      }
+
+      date_trunc = "date_trunc('day', library_events.created_at)"
+      progress0 = "(changed_data#>>'{progress,0}')::integer"
+      progress1 = "(changed_data#>>'{progress,1}')::integer"
+      progress = "(#{progress1} - #{progress0})"
+
+      case media_column
+      when :anime then watch_time = "sum(#{progress} * anime.episode_length)"
+      when :manga then watch_time = 0
+      end
+
+      activity_dates = user.library_events.where("changed_data ? 'progress'")
```

Comment:
You should add a comment explaining what this is building, what this beast is gonna return